src/python/zquantum/qcbm/target_thermal_states.py

```diff
@@ -0,0 +1,235 @@
+        for j in range(i, n_spins):
+            if i == j:
+                continue
+            two_body_couplings[i, j] = two_body_couplings[j, i] = np.random.normal(
+                0, np.sqrt(n_spins)
+            )
+    return external_fields, two_body_couplings
+
+
+def get_thermal_target_distribution_dict(
+    n_spins: int,
+    temperature: float,
+    hamiltonian_parameters: typing.List[np.array],  # TODO: add docstring
+) -> typing.Dict:
+    """Generates thermal states target distribution, saved in a dict where keys are bitstrings and
+    values are corresponding probabilities according to the Boltzmann Distribution formula.
+
+    Args:
+        n_spins: positive number of spins in the Ising system
+        temperature: temperature factor in the boltzman distribution
+        hamiltonian_parameters: values of hamiltonian parameters, namely external fields and two body couplings.
+
+    Returns:
+       Thermal target distribution.
+       Number of positive spins in the spin state. 
+    """
+    partition_function = 0
+    external_fields, two_body_couplings = hamiltonian_parameters
+    beta = 1.0 / temperature
+    distribution = {}
+
+    for spin in range(int(2 ** n_spins)):
+        ising_bitstring = int2ising(spin, n_spins)
+        energy = 0
+        for i in range(n_spins):
+            energy -= ising_bitstring[i] * external_fields[i]
+            for j in range(n_spins):
+                if j == i + 1:
+                    energy -= (
+                        ising_bitstring[i]
+                        * ising_bitstring[j]
+                        * two_body_couplings[i, j]
+                    )
+                else:
+                    continue
+        boltzmann_factor = np.exp(energy * beta)
+        partition_function += boltzmann_factor
+
+        binary_bitstring = convert_tuples_to_bitstrings([dec2bin(spin, n_spins)])[-1]
+        distribution[binary_bitstring] = boltzmann_factor
+
+    normalized_distribution = {
+        key: value / partition_function for key, value in distribution.items()
+    }
+
+    return normalized_distribution
+
+
+def get_target_bitstring_distribution(
+    n_spins: int,
+    temperature: float,
+    hamiltonian_parameters: typing.List[np.array],  # TODO: add docstring
+) -> BitstringDistribution:
+    """Generates thermal states target data as BitStringDistribution object
+
+    Args:
+        n_spins: positive number of spins in the Ising system
+        temperature: temperature factor in the boltzman distribution
+
+    Returns:
+       Thermal target Bistring Distribution.
+    """
+    return BitstringDistribution(
+        get_thermal_target_distribution_dict(
+            n_spins, temperature, hamiltonian_parameters
+        )
+    )
+
+
+def cumulate(
+    n_spins: int,
+    temperature: float,
+    hamiltonian_parameters: typing.List[np.array],  # TODO: add docstring
+) -> typing.List[float]:
+    """Generates array of cumulative probabilities from target_distribution.
```

Comment:
I don't totally understand this function. Is the idea to do something like the following:
`{"00": 0.2, "01": 0.15, "10": 0.6, "11": 0.05}` -> `[0.2, 0.35, 0.95, 1.0]`?

Reply:
If so, how do you know that the keys in the distribution dictionary are arranged in the proper order?

Reply:
It only works if you know that they are in the proper order, and for our case, we build the thermal state distribution to have the proper order, so it works. But for another random distribution, you would have to make sure it's in order first.

Reply:
Check if it's sorted and if not we sort it, using https://github.com/zapatacomputing/z-quantum-core/blob/4680301d4b43d4de0888456291189b191ba1f04c/src/python/zquantum/core/utils.py#L645 + https://github.com/zapatacomputing/z-quantum-qcbm/blob/2572fcb386ea7b0542b7290d0f504ce88b668858/examples/plot_qcbm_opt_history.py#L45

Reply:
Took out cumulate anyway as we switch to Manuel's function from z-quantum-core!